File: python-backend/rag_search.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel
import os
from dotenv import load_dotenv
from opensearchpy import OpenSearch
from rag_uploadfiles import generate_embeddings
from openai import OpenAI
from util import generateToken

generateToken()

router = APIRouter()
load_dotenv()

host = os.environ.get("OPENSEARCH_HOST")
port = os.environ.get("OPENSEARCH_PORT")
username = os.environ.get("OPENSEARCH_USERNAME")
password = os.environ.get("OPENSEARCH_PASSWORD")

header_name = os.getenv('GATEWAY_HEADER_NAME')
header_value = os.getenv('GATEWAY_HEADER_VALUE')
headers = {
    header_name: header_value
}
client = OpenAI(default_headers=headers)

# OpenSearch configuration
OPENSEARCH_CONFIG = {
    "hosts": [{"host": host, "port": port}],
    "http_auth": (username, password),
    "http_compress": True,
    "use_ssl": True,
    "verify_certs": False,
    "ssl_assert_hostname": False,
    "ssl_show_warn": False,
}
INDEX_NAME = "files"

class searchRequest(BaseModel):
    query: str

@router.post("/search")
async def search(request: searchRequest):
    query = request.query
    if not query:
        raise HTTPException(status_code=400, detail="Query cannot be empty")

    # Create embedding for the query
    try:
        query_embedding = generate_embeddings([query])[0]  # Get the first element of the list
        search_client = OpenSearch(**OPENSEARCH_CONFIG)
        search_body = {        
        "_source": ["content"],  # Only retrieve necessary fields
        "query": {
            "knn": {
                "embedding":{
                    "vector": query_embedding,
                    "k": 3,
                    }
                }
            }
        } 

        # Search for similar documents based on the query embedding
        response = search_client.search(index=INDEX_NAME, body=search_body)

        # Extract content and source from the documents
        documents_string = ''
        # # match_all query returns all documents, so we need to filter based on cosine similarity
        for hit in response["hits"]["hits"]:
            doc = hit["_source"]
            documents_string += doc['content'] + " "
        
        # Note: Match_all query returns all documents, so we need to filter based on cosine similarity
        
        # Challenge 3: Frame the prompt to include the query and the context in documents_string
        prompt = f"Context: {documents_string}"
        

        # Call OpenAI API
        response =  client.chat.completions.create(
        model="gpt-4o-2024-08-06",
        messages=[
            {"role": "system", "content": "You are a helpful assistant."},
            {"role": "user", "content": prompt},
        ],
        )
        #Challenge 4: Call the OpenAI API to get the response for the prompt

        response_str = response.choices[0].message.content
        return JSONResponse(content={"response": response_str})

    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to create embedding: {str(e)}"
        )

File: nodejs-backend-solved/routes/rag_search.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const fs = require('fs');
const path = require('path');
const os = require('os');
const { generateToken } = require('../util');
const { v4: uuidv4 } = require('uuid');
const { Client: OpenSearchClient } = require('@opensearch-project/opensearch');
const { OpenAI } = require('openai');
const { Document } = require("langchain/document");
const { OpenAIEmbeddings } = require("@langchain/openai");
const { OpenSearchVectorStore } = require("@langchain/community/vectorstores/opensearch");


require('dotenv').config();
(async () => {
  await generateToken();
})();

const header_name = process.env.GATEWAY_HEADER_NAME
const header_value = process.env.GATEWAY_HEADER_VALUE
const headers = {
    header_name: header_value,
};
const openAIClient = new OpenAI(headers);

const INDEX_NAME = 'files';

const OPENSEARCH_CONFIG = {
  node: `https://${process.env.OPENSEARCH_HOST}:${process.env.OPENSEARCH_PORT}`,
  auth: {
    username: process.env.OPENSEARCH_USERNAME,
    password: process.env.OPENSEARCH_PASSWORD,
  },
  ssl: {
    rejectUnauthorized: false,
  },
};

const opensearchClient = new OpenSearchClient(OPENSEARCH_CONFIG);

 async function generate_embeddings(texts) {
  const response = await openAIClient.embeddings.create({
    input: texts,
    model: 'text-embedding-3-large',
    dimensions: 256
  });
  return response.data.map(item => item.embedding);
}

// Endpoint to search
router.post('/search', async (req, res) => {
  const searchQuery = req.body.query;
  // Generate embeddings for the search query
  const query_embedding = await generate_embeddings([searchQuery]);
  
  search_body = {
    "size": 3, 
    "_source": [
        "name",
        "content"
      
    ],  
    "query": {
   "script_score": {
     "query": {
       "match_all": {}
     },
     "script": {
       "source": "knn_score",
       "lang": "knn",
       "params": {
         "field": "embedding",
         "query_value": query_embedding[0],
         "space_type": "cosinesimil",
         "k": 3
       }
     }
   }
 }
  }


  // Search for similar documents based on the query embedding
//Search for similar documents based on the query embedding
  response = await opensearchClient.search({
    index: INDEX_NAME, 
    body: search_body
  })
  // Extract documents and their embeddings
  let documents_string = "";
  // match_all query returns all documents, so we need to filter based on cosine similarity
  response.body.hits.hits.forEach(hit => {
    const doc = hit._source;
    documents_string += doc.content;
    documents_string += `\nSource: ${doc.name}\n\n`;
  });

  // OpenAI call
  const prompt = `Answer the question: ${searchQuery} based only on the following context. At the end also mention source from context:
  context:  ${documents_string}`;

  // Call OpenAI API
  const openaiResponse = await openAIClient.chat.completions.create({
    model: "gpt-4o-2024-05-13",
    temperature: 1,  // Higher temperature can result in more creative responses apart from context 
    messages: [
      {
        role: "system",
        content: "You are an AI assistant tasked with answering questions using the provided context as the primary source of information.",
      },
      { role: "user", content: prompt },
    ],
  });

  const questions_text = openaiResponse.choices[0].message.content;
  res.send({ response: questions_text });

  res.send({ message: `Search results for: ${searchQuery}` });
});

module.exports = router;
